cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+};
+
 const FunctionDoc sort_indices_doc(
+    "Return the indices that would sort an array, record batch or table",
+    ("This function computes an array of indices that define a stable sort\n"
+     "of the input array, record batch or table.  Null values are considered\n"
+     "greater than any other value and are therefore sorted at the end of the\n"
+     "input. For floating-point types, NaNs are considered greater than any\n"
+     "other non-null value, but smaller than null values."),
+    {"input"}, "SortOptions");
+
+class SortIndicesMetaFunction : public MetaFunction {
+ public:
+  SortIndicesMetaFunction()
+      : MetaFunction("sort_indices", Arity::Unary(), &sort_indices_doc) {}
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    const SortOptions& sort_options = static_cast<const SortOptions&>(*options);
+    switch (args[0].kind()) {
+      case Datum::ARRAY:
+        return SortIndices(*args[0].make_array(), sort_options, ctx);
+        break;
+      case Datum::CHUNKED_ARRAY:
+        return SortIndices(*args[0].chunked_array(), sort_options, ctx);
+        break;
+      case Datum::RECORD_BATCH: {
+        ARROW_ASSIGN_OR_RAISE(auto table,
+                              Table::FromRecordBatches({args[0].record_batch()}));
+        return SortIndices(*table, sort_options, ctx);
+      } break;
+      case Datum::TABLE:
+        return SortIndices(*args[0].table(), sort_options, ctx);
+        break;
+      default:
+        break;
+    }
+    return Status::NotImplemented(
+        "Unsupported types for sort_indices operation: "
+        "values=",
+        args[0].ToString());
+  }
+
+ private:
+  Result<std::shared_ptr<Array>> SortIndices(const Array& values,
+                                             const SortOptions& options,
+                                             ExecContext* ctx) const {
+    SortOrder order = SortOrder::ASCENDING;
+    if (!options.sort_keys.empty()) {
+      order = options.sort_keys[0].order;
+    }
+    ArraySortOptions array_options(order);
+    ARROW_ASSIGN_OR_RAISE(
+        Datum result, CallFunction("array_sort_indices", {values}, &array_options, ctx));
+    return result.make_array();
+  }
+
+  Result<std::shared_ptr<Array>> SortIndices(const ChunkedArray& values,
+                                             const SortOptions& options,
+                                             ExecContext* ctx) const {
+    SortOrder order = SortOrder::ASCENDING;
+    if (!options.sort_keys.empty()) {
+      order = options.sort_keys[0].order;
+    }
+    ArraySortOptions array_options(order);
+
+    std::shared_ptr<Array> array_values;
+    if (values.num_chunks() == 1) {
+      array_values = values.chunk(0);
+    } else {
+      ARROW_ASSIGN_OR_RAISE(array_values,
+                            Concatenate(values.chunks(), ctx->memory_pool()));
```

Comment:
This is weird. Why are you concatenating chunks for the ChunkedArray case but not for the Table case? I would expect both cases to be handled similarly (regardless of the solution).

Reply:
Ah, I forgot to mention this.
I wanted to make ChunkedArray case as follow-up task. I wanted to focus to Table in this pull request.

But I've implemented naive ChunkedArray support that doesn't use `Concatenate()` in this pull request. It sorts each chunk and then merge them but it doesn't use threads.